fvm/evm/offchain/query/view.go

```diff
@@ -0,0 +1,371 @@
+	bv, err := state.NewBaseView(v.storage, v.rootAddr)
+	if err != nil {
+		return gethCommon.Hash{}, err
+	}
+	return bv.GetCodeHash(addr)
+}
+
+// GetSlab returns the slab for the given address and key
+// can be used for the `eth_getStorageAt` endpoint
+func (v *View) GetSlab(addr gethCommon.Address, key gethCommon.Hash) (gethCommon.Hash, error) {
+	bv, err := state.NewBaseView(v.storage, v.rootAddr)
+	if err != nil {
+		return gethCommon.Hash{}, err
+	}
+	return bv.GetState(types.SlotAddress{
+		Address: addr,
+		Key:     key,
+	})
+}
+
+// DryCall runs a call off-chain and returns the results
+// accepts override storage and precompiled call options
+// as well as custom tracer.
+func (v *View) DryCall(
+	from gethCommon.Address,
+	to gethCommon.Address,
+	data []byte,
+	value *big.Int,
+	gasLimit uint64,
+	gasPrice *big.Int,
+	opts ...DryCallOption,
+) (*types.Result, error) {
+	// apply all the options
+	for _, op := range opts {
+		err := op(v)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	blks, err := blocks.NewBlocks(v.chainID, v.rootAddr, v.storage)
+	if err != nil {
+		return nil, err
+	}
+
+	// create context
+	ctx, err := sync.CreateBlockContext(v.chainID, blks, v.tracer)
+	if err != nil {
+		return nil, err
+	}
+	ctx.ExtraPrecompiledContracts = v.extraPCs
+
+	// create emulator
+	em := emulator.NewEmulator(v.storage, v.rootAddr)
+
+	// create a new block view
+	bv, err := em.NewBlockView(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if gasLimit > v.maxCallGasLimit {
+		return nil, fmt.Errorf(
+			"gas limit is bigger than max gas limit allowed %d > %d",
+			gasLimit, v.maxCallGasLimit,
+		)
+	}
+
+	res, err := bv.DirectCall(
+		&types.DirectCall{
+			From:     types.NewAddress(from),
+			To:       types.NewAddress(to),
+			Data:     data,
+			Value:    value,
+			GasLimit: gasLimit,
+		},
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return res, nil
+}
+
+// DryRunOption captures a options
+// to be applied before the execution of a dry call.
+type DryCallOption func(v *View) error
+
+// WithStateOverrideBalance constructs a dry call option
+// that replaces the balance of an address before the execution a dry call.
+func WithStateOverrideBalance(
+	addr gethCommon.Address,
+	balance *big.Int,
+) DryCallOption {
+	return func(v *View) error {
+		baseView, err := state.NewBaseView(v.storage, v.rootAddr)
+		if err != nil {
+			return err
+		}
+		nonce, err := baseView.GetNonce(addr)
+		if err != nil {
+			return err
+		}
+		code, err := baseView.GetCode(addr)
+		if err != nil {
+			return err
+		}
+		codeHash, err := baseView.GetCodeHash(addr)
+		if err != nil {
+			return err
+		}
+
+		convertedBalance, overflow := uint256.FromBig(balance)
+		if overflow {
+			return errors.New("balance too large")
+		}
+
+		err = baseView.UpdateAccount(addr, convertedBalance, nonce, code, codeHash)
+		if err != nil {
+			return err
+		}
+		return baseView.Commit()
```

Comment:
I must miss something here. 

Does the `Commit` call persist the balance update into database?

If we want to override the balance for making a dry call, do we also need to revert the balance update after the dry call? Because otherwise, the dry call isn't "dry" since the database is updated, it will affect the next dry call. No?

Reply:
storage that is passed to the baseView is of type storage.EphemeralStorage. so all the changes would be discarded.